libvast/builtins/operators/head.cpp

```diff
@@ -0,0 +1,98 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <vast/concept/parseable/numeric/integral.hpp>
+#include <vast/concept/parseable/vast/pipeline.hpp>
+#include <vast/error.hpp>
+#include <vast/logger.hpp>
+#include <vast/pipeline.hpp>
+#include <vast/plugin.hpp>
+#include <vast/table_slice.hpp>
+
+#include <arrow/type.h>
+
+namespace vast::plugins::head {
+
+namespace {
+
+class head_operator : public pipeline_operator {
+public:
+  explicit head_operator(uint64_t remaining) noexcept : remaining_{remaining} {
+    // nop
+  }
+
+  caf::error add(table_slice slice) override {
+    // When switching to the new operator API, we ideally want to implement an
+    // early exit for head here; for now this operator ends up being a lot of
+    // no-ops once we reached the limit.
+    if (remaining_ == 0)
+      return {};
+    slice = vast::head(slice, remaining_);
+    remaining_ -= slice.rows();
```

Comment:
Maybe add an assert that `remaining_` doesn't underflow here?

Reply:
It's safe and checked for, but I suppose an extra assertion for the reader of this code doesn't hurt.